File: BikeDataOverlayMaui/Services/BluetoothService.cs
using Plugin.BLE;
using Plugin.BLE.Abstractions.Contracts;
using Plugin.BLE.Abstractions.EventArgs;
using Plugin.BLE.Abstractions.Exceptions;
using BikeDataOverlayMaui.Models;
using System.Collections.ObjectModel;

namespace BikeDataOverlayMaui.Services;

public class BluetoothService : IBluetoothService
{
    private readonly IBluetoothLE _bluetoothLE;
    private readonly List<BluetoothDevice> _connectedDevices = new();
    private readonly List<ICharacteristic> _notificationCharacteristics = new();
    private CyclingMetrics _currentMetrics = new();
    private bool _isConnected = false;
    private string? _connectedDeviceName;
    
    // FTMS Service UUIDs
    private readonly Guid FitnessMachineServiceUuid = new("00001826-0000-1000-8000-00805f9b34fb");
    private readonly Guid IndoorBikeDataCharacteristicUuid = new("00002ad2-0000-1000-8000-00805f9b34fb");
    private readonly Guid HeartRateServiceUuid = new("0000180d-0000-1000-8000-00805f9b34fb");
    private readonly Guid HeartRateCharacteristicUuid = new("00002a37-0000-1000-8000-00805f9b34fb");
    
    public event EventHandler<CyclingMetrics>? MetricsUpdated;
    public event EventHandler<bool>? ConnectionStatusChanged;
    
    public bool IsConnected => _isConnected;
    public string? ConnectedDeviceName => _connectedDeviceName;
    public IReadOnlyList<BluetoothDevice> ConnectedDevices => _connectedDevices.AsReadOnly();

    public BluetoothService()
    {
        _bluetoothLE = CrossBluetoothLE.Current;
        _currentMetrics = new CyclingMetrics
        {
            Watts = 0,
            Cadence = 0,
            Speed = 0,
            HeartRate = 0,
            Timestamp = DateTime.UtcNow
        };
    }

    public async Task<bool> ScanAndConnectAsync(CancellationToken cancellationToken = default)
    {
        try
        {
            Console.WriteLine("=== MAUI Bluetooth FTMS Trainer Connection ===");
            
            if (_isConnected)
            {
                Console.WriteLine("Already connected to a trainer");
                return true;
            }

            await DisconnectAsync();

            Console.WriteLine("Scanning for FTMS devices...");
            var devices = await ScanForDevicesAsync(TimeSpan.FromSeconds(10));
            
            if (!devices.Any())
            {
                Console.WriteLine("No FTMS devices found. Starting simulation mode.");
                return await StartSimulationMode(cancellationToken);
            }

            // Try to connect to the first available device
            var device = devices.First();
            var connected = await ConnectToDeviceAsync(device.Id, cancellationToken);
            
            if (!connected)
            {
                Console.WriteLine("Failed to connect to any devices. Starting simulation mode.");
                return await StartSimulationMode(cancellationToken);
            }

            return true;
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Real connection failed: {ex.Message}");
            Console.WriteLine("Starting simulation mode.");
            return await StartSimulationMode(cancellationToken);
        }
    }

    public async Task<List<BluetoothDevice>> ScanForDevicesAsync(TimeSpan? timeout = null)
    {
        var devices = new List<BluetoothDevice>();
        
        try
        {
            Console.WriteLine("Starting Bluetooth device scan...");
            
            var state = _bluetoothLE.State;
            if (state != BluetoothState.On)
            {
                Console.WriteLine($"Bluetooth is not on. Current state: {state}");
                return devices;
            }

            var adapter = _bluetoothLE.Adapter;
            var scanTimeout = timeout ?? TimeSpan.FromSeconds(10);

            var scanResults = new List<Plugin.BLE.Abstractions.Contracts.IDevice>();
            
            adapter.DeviceDiscovered += (s, e) =>
            {
                var device = e.Device;
                if (!string.IsNullOrEmpty(device.Name) && 
                    (device.Name.ToLower().Contains("trainer") || 
                     device.Name.ToLower().Contains("bike") ||
                     device.Name.ToLower().Contains("kickr") ||
                     device.Name.ToLower().Contains("zwift")))
                {
                    scanResults.Add(device);
                    Console.WriteLine($"Device discovered: {device.Name} (RSSI: {device.Rssi})");
                }
            };

            await adapter.StartScanningForDevicesAsync();
            
            // Wait for scan timeout
            await Task.Delay(scanTimeout);
            
            // Stop scanning
            await adapter.StopScanningForDevicesAsync();

            foreach (var result in scanResults)
            {
                devices.Add(new BluetoothDevice
                {
                    Id = result.Id.ToString(),
                    Name = result.Name ?? "Unknown Device",
                    DeviceType = "Fitness Device",
                    IsConnected = result.State == Plugin.BLE.Abstractions.DeviceState.Connected,
                    CanConnect = true,
                    Status = result.State == Plugin.BLE.Abstractions.DeviceState.Connected ? "Connected" : "Available",
                    LastSeen = DateTime.UtcNow,
                    DeviceInfo = new Models.DeviceInfo
                    {
                        Manufacturer = "Unknown",
                        Model = result.Name ?? "Unknown",
                        Type = "Fitness Device"
                    }
                });
            }
            
            Console.WriteLine($"Bluetooth scan finished. Found {devices.Count} fitness devices.");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error during Bluetooth scan: {ex.Message}");
        }
        
        return devices;
    }

    public async Task<bool> ConnectToDeviceAsync(string deviceId, CancellationToken cancellationToken = default)
    {
        try
        {
            var adapter = _bluetoothLE.Adapter;
            var deviceGuid = Guid.Parse(deviceId);
            
            Console.WriteLine($"Attempting to connect to device: {deviceId}");
            
            var device = await adapter.ConnectToKnownDeviceAsync(deviceGuid, cancellationToken: cancellationToken);
            
            if (device == null || device.State != Plugin.BLE.Abstractions.DeviceState.Connected)
            {
                Console.WriteLine("Failed to connect to device");
                return false;
            }

            Console.WriteLine($"Successfully connected to: {device.Name}");
            
            // Get FTMS service
            var services = await device.GetServicesAsync();
            var ftmsService = services.FirstOrDefault(s => s.Id == FitnessMachineServiceUuid);
            
            if (ftmsService != null)
            {
                var characteristics = await ftmsService.GetCharacteristicsAsync();
                var indoorBikeChar = characteristics.FirstOrDefault(c => c.Id == IndoorBikeDataCharacteristicUuid);
                
                if (indoorBikeChar != null && indoorBikeChar.CanRead)
                {
                    indoorBikeChar.ValueUpdated += OnIndoorBikeDataReceived;
                    await indoorBikeChar.StartUpdatesAsync();
                    _notificationCharacteristics.Add(indoorBikeChar);
                    Console.WriteLine("Started notifications for Indoor Bike Data");
                }
            }

            // Get Heart Rate service
            var hrService = services.FirstOrDefault(s => s.Id == HeartRateServiceUuid);
            if (hrService != null)
            {
                var characteristics = await hrService.GetCharacteristicsAsync();
                var hrChar = characteristics.FirstOrDefault(c => c.Id == HeartRateCharacteristicUuid);
                
                if (hrChar != null && hrChar.CanRead)
                {
                    hrChar.ValueUpdated += OnHeartRateDataReceived;
                    await hrChar.StartUpdatesAsync();
                    _notificationCharacteristics.Add(hrChar);
                    Console.WriteLine("Started notifications for Heart Rate");
                }
            }

            var bluetoothDevice = new BluetoothDevice
            {
                Id = deviceId,
                Name = device.Name ?? "Unknown Device",
                DeviceType = "Fitness Device",
                IsConnected = true,
                Status = "Connected",
                LastSeen = DateTime.UtcNow
            };

            _connectedDevices.Add(bluetoothDevice);
            _connectedDeviceName = device.Name;
            _isConnected = true;
            
            ConnectionStatusChanged?.Invoke(this, true);
            Console.WriteLine($"Successfully connected to: {device.Name}");
            
            return true;
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error connecting to device {deviceId}: {ex.Message}");
            return false;
        }
    }

    public async Task DisconnectAsync()
    {
        try
        {
            if (_isConnected)
            {
                Console.WriteLine($"Disconnecting from devices: {_connectedDeviceName}");
                
                // Stop all notifications
                foreach (var characteristic in _notificationCharacteristics)
                {
                    try
                    {
                        await characteristic.StopUpdatesAsync();
                    }
                    catch (Exception ex)
                    {
                        Console.WriteLine($"Error stopping updates for characteristic: {ex.Message}");
                    }
                }
                _notificationCharacteristics.Clear();

                var adapter = _bluetoothLE.Adapter;
                foreach (var device in adapter.ConnectedDevices)
                {
                    await adapter.DisconnectDeviceAsync(device);
                    Console.WriteLine($"Disconnected from {device.Name}");
                }
                
                _connectedDevices.Clear();
                _isConnected = false;
                _connectedDeviceName = null;
                
                ConnectionStatusChanged?.Invoke(this, false);
                Console.WriteLine("All devices disconnected successfully");
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Disconnect error: {ex.Message}");
        }
    }

    public CyclingMetrics GetCurrentMetrics()
    {
        return new CyclingMetrics
        {
            Watts = _currentMetrics.Watts,
            Cadence = _currentMetrics.Cadence,
            Speed = _currentMetrics.Speed,
            HeartRate = _currentMetrics.HeartRate,
            Timestamp = _currentMetrics.Timestamp
        };
    }

    private Task<bool> StartSimulationMode(CancellationToken cancellationToken = default)
    {
        try
        {
            Console.WriteLine("Starting FTMS simulation mode...");
            
            _connectedDeviceName = "Simulated KICKR CORE";
            Console.WriteLine($"Simulation Device: {_connectedDeviceName}");
            
            // Start receiving simulated FTMS data
            StartFtmsDataSimulation();
            
            _isConnected = true;
            ConnectionStatusChanged?.Invoke(this, true);
            Console.WriteLine($"Successfully started simulation mode: {_connectedDeviceName}");
            
            return Task.FromResult(true);
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Simulation mode failed: {ex.Message}");
            _isConnected = false;
            _connectedDeviceName = null;
            ConnectionStatusChanged?.Invoke(this, false);
            return Task.FromResult(false);
        }
    }

    private void StartFtmsDataSimulation()
    {
        Task.Run(async () =>
        {
            var random = new Random();
            var elapsedSeconds = 0;
            
            Console.WriteLine("Starting FTMS Indoor Bike Data simulation...");
            
            while (_isConnected)
            {
                try
                {
                    // Simulate realistic cycling workout data with variations
                    var workoutIntensity = 0.7 + 0.3 * Math.Sin(elapsedSeconds / 30.0);
                    
                    var basePower = 180 + (int)(50 * workoutIntensity);
                    var baseCadence = 85 + (int)(15 * workoutIntensity);
                    var baseSpeed = 28.0 + (8.0 * workoutIntensity);
                    
                    // Add realistic variations
                    var powerVariation = (int)(Math.Sin(elapsedSeconds / 10.0) * 15);
                    var cadenceVariation = (int)(Math.Cos(elapsedSeconds / 12.0) * 8);
                    var speedVariation = Math.Sin(elapsedSeconds / 15.0) * 3.0;
                    
                    _currentMetrics = new CyclingMetrics
                    {
                        Watts = Math.Max(0, basePower + powerVariation + random.Next(-8, 8)),
                        Cadence = Math.Max(0, baseCadence + cadenceVariation + random.Next(-3, 3)),
                        Speed = Math.Max(0, baseSpeed + speedVariation + random.NextDouble() * 2 - 1),
                        HeartRate = 135 + (int)(25 * workoutIntensity) + random.Next(-5, 8),
                        Timestamp = DateTime.UtcNow
                    };
                    
                    elapsedSeconds++;
                    MetricsUpdated?.Invoke(this, _currentMetrics);
                    
                    if (elapsedSeconds % 10 == 0)
                    {
                        Console.WriteLine($"Simulated Data - Power: {_currentMetrics.Watts}W, " +
                                        $"Cadence: {_currentMetrics.Cadence}RPM, " +
                                        $"Speed: {_currentMetrics.Speed:F1}km/h, " +
                                        $"HR: {_currentMetrics.HeartRate}BPM");
                    }
                    
                    await Task.Delay(1000);
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error in FTMS data simulation: {ex.Message}");
                }
            }
            
            Console.WriteLine("FTMS data simulation stopped");
        });
    }

    private void OnIndoorBikeDataReceived(object? sender, CharacteristicUpdatedEventArgs e)
    {
        try
        {
            var data = e.Characteristic.Value;
            if (data != null && data.Length >= 4)
            {
                // Parse FTMS Indoor Bike Data (simplified)
                var flags = BitConverter.ToUInt16(data, 0);
                var dataIndex = 2;
                
                if ((flags & 0x01) != 0 && data.Length >= dataIndex + 2)
                {
                    _currentMetrics.Speed = BitConverter.ToUInt16(data, dataIndex) / 100.0;
                    dataIndex += 2;
                }
                
                if ((flags & 0x02) != 0 && data.Length >= dataIndex + 2)
                {
                    _currentMetrics.Cadence = BitConverter.ToUInt16(data, dataIndex) / 2.0;
                    dataIndex += 2;
                }
                
                if ((flags & 0x04) != 0 && data.Length >= dataIndex + 2)
                {
                    _currentMetrics.Watts = BitConverter.ToUInt16(data, dataIndex);
                    dataIndex += 2;
                }
                
                _currentMetrics.Timestamp = DateTime.UtcNow;
                MetricsUpdated?.Invoke(this, _currentMetrics);
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error processing indoor bike data: {ex.Message}");
        }
    }

    private void OnHeartRateDataReceived(object? sender, CharacteristicUpdatedEventArgs e)
    {
        try
        {
            var data = e.Characteristic.Value;
            if (data != null && data.Length >= 2)
            {
                // Parse Heart Rate data
                var flags = data[0];
                var heartRate = (flags & 0x01) == 0 ? data[1] : BitConverter.ToUInt16(data, 1);
                
                _currentMetrics.HeartRate = heartRate;
                _currentMetrics.Timestamp = DateTime.UtcNow;
                MetricsUpdated?.Invoke(this, _currentMetrics);
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error processing heart rate data: {ex.Message}");
        }
    }

    public void Dispose()
    {
        try
        {
            if (_isConnected)
            {
                DisconnectAsync().Wait(TimeSpan.FromSeconds(2));
            }
            
            _connectedDevices.Clear();
            _notificationCharacteristics.Clear();
            _isConnected = false;
            Console.WriteLine("BluetoothService disposed");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error during BluetoothService disposal: {ex.Message}");
        }
    }
}

File: frontend/src/services/BluetoothService.ts
import noble from '@stoprocent/noble';
import { EventEmitter } from 'events';
import { FTMSService } from './FTMSService';
import { CyclingMetrics, BluetoothDevice } from '../types/CyclingMetrics';

interface NoblePeripheral {
  id: string;
  advertisement: {
    localName?: string;
    shortLocalName?: string;
    serviceUuids: string[];
  };
  state: string;
  connect(callback: (error?: Error) => void): void;
  disconnect(callback?: (error?: Error) => void): void;
  discoverServices(serviceUuids: string[], callback: (error?: Error, services?: any[]) => void): void;
  on(event: string, listener: (...args: any[]) => void): void;
}

interface DiscoveredDevice {
  id: string;
  name: string;
  peripheral: NoblePeripheral;
  isConnected: boolean;
  deviceInfo: any;
  services: string[];
}

export class BluetoothService extends EventEmitter {
  private isConnected = false;
  private connectedDeviceName: string | null = null;
  private connectedDevices = new Map<string, DiscoveredDevice>();
  private discoveredDevices = new Map<string, DiscoveredDevice>();
  private currentMetrics: CyclingMetrics = {
    watts: 0,
    cadence: 0,
    speed: 0,
    heartRate: 0,
    timestamp: new Date().toISOString()
  };
  
  private ftmsService: FTMSService;
  private isScanning = false;
  private scanTimeout: NodeJS.Timeout | null = null;
  private simulationInterval: NodeJS.Timeout | null = null;

  constructor() {
    super();
    
    this.ftmsService = new FTMSService();
    this.ftmsService.on('metricsUpdate', (metrics) => {
      this.currentMetrics = {
        ...metrics,
        timestamp: new Date().toISOString()
      };
      this.emit('metricsUpdate', this.currentMetrics);
    });
    
    this.initializeNoble();
  }

  private initializeNoble(): void {
    noble.on('stateChange', (state: string) => {
      console.log('Bluetooth state changed to:', state);
      if (state === 'poweredOn') {
        console.log('Bluetooth is powered on and ready');
      } else {
        console.log('Bluetooth is not available:', state);
        if (this.isConnected) {
          this.disconnect();
        }
      }
    });

    noble.on('discover', (peripheral: NoblePeripheral) => {
      this.handleDeviceDiscovered(peripheral);
    });

    noble.on('scanStart', () => {
      console.log('Bluetooth scan started');
      this.isScanning = true;
    });

    noble.on('scanStop', () => {
      console.log('Bluetooth scan stopped');
      this.isScanning = false;
    });
  }

  private handleDeviceDiscovered(peripheral: NoblePeripheral): void {
    const hasHeartRateService = peripheral.advertisement.serviceUuids.includes('180d');
    const hasCyclingPowerService = peripheral.advertisement.serviceUuids.includes('1818');
    const hasFTMSService = peripheral.advertisement.serviceUuids.includes('1826');
    const hasCyclingSpeedCadenceService = peripheral.advertisement.serviceUuids.includes('1816');
    
    if (hasHeartRateService || hasCyclingPowerService || hasFTMSService || hasCyclingSpeedCadenceService) {
      const deviceName = peripheral.advertisement.localName || peripheral.advertisement.shortLocalName || `Unknown Device (${peripheral.id})`;
      
      console.log(`Discovered fitness device: ${deviceName} (${peripheral.id})`);
      console.log(`  Services: ${peripheral.advertisement.serviceUuids.join(', ')}`);
      
      const device: DiscoveredDevice = {
        id: peripheral.id,
        name: deviceName,
        peripheral: peripheral,
        isConnected: false,
        deviceInfo: null,
        services: peripheral.advertisement.serviceUuids
      };
      
      this.discoveredDevices.set(peripheral.id, device);
    }
  }

  async scanForDevices(timeoutMs: number = 15000): Promise<BluetoothDevice[]> {
    return new Promise((resolve, reject) => {
      if ((noble as any).state !== 'poweredOn') {
        reject(new Error(`Bluetooth not ready. State: ${(noble as any).state}`));
        return;
      }

      if (this.isScanning) {
        console.log('Scan already in progress, stopping current scan...');
        noble.stopScanning();
      }

      this.discoveredDevices.clear();

      console.log(`Starting Bluetooth scan for ${timeoutMs / 1000} seconds...`);

      const serviceUuids = [
        '180d', // Heart Rate Service
        '1818', // Cycling Power Service
        '1826', // Fitness Machine Service (FTMS)
        '1816'  // Cycling Speed and Cadence Service
      ];

      noble.startScanning(serviceUuids, false);

      this.scanTimeout = setTimeout(() => {
        noble.stopScanning();
        const devices = Array.from(this.discoveredDevices.values()).map(device => ({
          id: device.id,
          name: device.name,
          isConnected: device.isConnected,
          deviceInfo: device.deviceInfo,
          services: device.services
        }));
        console.log(`Scan completed. Found ${devices.length} fitness devices.`);
        resolve(devices);
      }, timeoutMs);
    });
  }

  async connectToDevice(deviceId: string): Promise<boolean> {
    try {
      const device = this.discoveredDevices.get(deviceId);
      if (!device) {
        console.log(`Device ${deviceId} not found, scanning...`);
        await this.scanForDevices(5000);
        const deviceAfterScan = this.discoveredDevices.get(deviceId);
        if (!deviceAfterScan) {
          throw new Error(`Device ${deviceId} not found`);
        }
        return this.connectToDevice(deviceId);
      }

      if (this.connectedDevices.has(deviceId)) {
        console.log(`Device ${device.name} is already connected`);
        return true;
      }

      console.log(`Connecting to device: ${device.name} (${deviceId})`);

      return new Promise((resolve, reject) => {
        const connectTimeout = setTimeout(() => {
          reject(new Error('Connection timeout'));
        }, 10000);

        device.peripheral.connect((error) => {
          clearTimeout(connectTimeout);
          
          if (error) {
            console.error(`Failed to connect to ${device.name}:`, error);
            reject(error);
            return;
          }

          console.log(`Successfully connected to ${device.name}`);
          
          device.isConnected = true;
          this.connectedDevices.set(deviceId, device);
          this.updateConnectionStatus();

          device.peripheral.on('disconnect', () => {
            console.log(`Device ${device.name} disconnected`);
            device.isConnected = false;
            this.connectedDevices.delete(deviceId);
            this.updateConnectionStatus();
          });

          this.setupDeviceServices(device).then(() => {
            resolve(true);
          }).catch((serviceError) => {
            console.error(`Failed to setup services for ${device.name}:`, serviceError);
            resolve(true);
          });
        });
      });
    } catch (error) {
      console.error(`Error connecting to device ${deviceId}:`, error);
      throw error;
    }
  }

  private async setupDeviceServices(device: DiscoveredDevice): Promise<void> {
    return new Promise((resolve, reject) => {
      device.peripheral.discoverServices([], (error: any, services?: any[]) => {
        if (error) {
          reject(error);
          return;
        }
        
        if (!services) {
          resolve();
          return;
        }

        console.log(`Discovered ${services.length} services for ${device.name}`);
        
        const ftmsService = services.find(service => 
          service.uuid === '1826' || service.uuid.toLowerCase().includes('1826')
        );

        if (ftmsService) {
          console.log(`Found FTMS service on ${device.name}`);
          this.setupFTMSService(device, ftmsService).then(resolve).catch(reject);
        } else {
          const powerService = services.find(service => service.uuid === '1818');
          const heartRateService = services.find(service => service.uuid === '180d');
          const speedCadenceService = services.find(service => service.uuid === '1816');
          
          if (powerService || heartRateService || speedCadenceService) {
            console.log(`Found fitness services on ${device.name}, setting up basic monitoring`);
            resolve();
          } else {
            console.log(`No recognized fitness services found on ${device.name}`);
            resolve();
          }
        }
      });
    });
  }

  private async setupFTMSService(device: DiscoveredDevice, ftmsService: any): Promise<void> {
    return new Promise((resolve, reject) => {
      ftmsService.discoverCharacteristics([], (error: any, characteristics: any[]) => {
        if (error) {
          reject(error);
          return;
        }

        console.log(`Found ${characteristics.length} FTMS characteristics for ${device.name}`);
        
        const indoorBikeDataChar = characteristics.find(char => 
          char.uuid === '2ad2' || char.uuid.toLowerCase().includes('2ad2')
        );

        if (indoorBikeDataChar && indoorBikeDataChar.properties.includes('notify')) {
          console.log(`Setting up Indoor Bike Data notifications for ${device.name}`);
          
          indoorBikeDataChar.on('data', (data: Buffer) => {
            const metrics = this.ftmsService.parseIndoorBikeData(data);
            if (metrics) {
              this.currentMetrics = {
                ...metrics,
                timestamp: new Date().toISOString()
              };
              this.emit('metricsUpdate', this.currentMetrics);
              console.log(`FTMS Data from ${device.name}:`, metrics);
            }
          });

          indoorBikeDataChar.subscribe((subscribeError: any) => {
            if (subscribeError) {
              console.error(`Failed to subscribe to FTMS notifications:`, subscribeError);
              reject(subscribeError);
            } else {
              console.log(`Successfully subscribed to FTMS notifications for ${device.name}`);
              resolve();
            }
          });
        } else {
          console.log(`Indoor Bike Data characteristic not found or not notifiable on ${device.name}`);
          resolve();
        }
      });
    });
  }

  private updateConnectionStatus(): void {
    this.isConnected = this.connectedDevices.size > 0;
    if (this.isConnected) {
      const deviceNames = Array.from(this.connectedDevices.values()).map(d => d.name);
      this.connectedDeviceName = deviceNames.join(', ');
    } else {
      this.connectedDeviceName = null;
    }
    this.emit('connectionStatusChanged', {
      isConnected: this.isConnected,
      deviceName: this.connectedDeviceName
    });
  }

  async scanAndConnect(): Promise<boolean> {
    try {
      console.log('=== Node.js Bluetooth FTMS Trainer Connection ===');
      
      if (this.isConnected) {
        console.log('Already connected to a trainer');
        return true;
      }

      console.log('Scanning for FTMS devices...');
      const devices = await this.scanForDevices(10000);
      
      if (devices.length === 0) {
        console.log('No FTMS devices found. Starting simulation mode.');
        return this.startSimulationMode();
      }

      const device = devices[0];
      console.log(`Attempting to connect to: ${device.name}`);
      
      const connected = await this.connectToDevice(device.id);
      if (connected) {
        console.log(`Successfully connected to: ${device.name}`);
        return true;
      } else {
        console.log('Failed to connect to device. Starting simulation mode.');
        return this.startSimulationMode();
      }
    } catch (error: any) {
      console.error('Real connection failed:', error.message);
      console.log('Starting simulation mode.');
      return this.startSimulationMode();
    }
  }

  private startSimulationMode(): boolean {
    console.log('Starting FTMS simulation mode...');
    
    this.connectedDeviceName = 'Simulated KICKR CORE';
    this.isConnected = true;
    this.updateConnectionStatus();
    
    this.simulationInterval = setInterval(() => {
      const now = Date.now();
      const elapsedSeconds = Math.floor(now / 1000) % 3600;
      
      const workoutIntensity = 0.7 + 0.3 * Math.sin(elapsedSeconds / 30.0);
      const basePower = 180 + Math.floor(50 * workoutIntensity);
      const baseCadence = 85 + Math.floor(15 * workoutIntensity);
      const baseSpeed = 28.0 + (8.0 * workoutIntensity);
      
      const powerVariation = Math.floor(Math.sin(elapsedSeconds / 10.0) * 15);
      const cadenceVariation = Math.floor(Math.cos(elapsedSeconds / 12.0) * 8);
      const speedVariation = Math.sin(elapsedSeconds / 15.0) * 3.0;
      
      this.currentMetrics = {
        watts: Math.max(0, basePower + powerVariation + Math.floor(Math.random() * 16 - 8)),
        cadence: Math.max(0, baseCadence + cadenceVariation + Math.floor(Math.random() * 6 - 3)),
        speed: Math.max(0, baseSpeed + speedVariation + (Math.random() * 2 - 1)),
        heartRate: 135 + Math.floor(25 * workoutIntensity) + Math.floor(Math.random() * 13 - 5),
        timestamp: new Date().toISOString()
      };
      
      this.emit('metricsUpdate', this.currentMetrics);
      
      if (elapsedSeconds % 10 === 0) {
        console.log(`Simulation Data - Power: ${this.currentMetrics.watts}W, ` +
                   `Cadence: ${this.currentMetrics.cadence}RPM, ` +
                   `Speed: ${this.currentMetrics.speed.toFixed(1)}km/h, ` +
                   `HR: ${this.currentMetrics.heartRate}BPM`);
      }
    }, 1000);
    
    console.log('Successfully started simulation mode');
    return true;
  }

  async disconnect(): Promise<void> {
    try {
      console.log('Disconnecting from all devices...');
      
      if (this.simulationInterval) {
        clearInterval(this.simulationInterval);
        this.simulationInterval = null;
      }

      if (this.isScanning) {
        noble.stopScanning();
      }

      for (const [deviceId, device] of this.connectedDevices) {
        try {
          if (device.peripheral && device.peripheral.state === 'connected') {
            console.log(`Disconnecting from ${device.name}`);
            device.peripheral.disconnect();
          }
        } catch (error) {
          console.error(`Error disconnecting from ${device.name}:`, error);
        }
      }

      this.connectedDevices.clear();
      this.isConnected = false;
      this.connectedDeviceName = null;
      
      this.currentMetrics = {
        watts: 0,
        cadence: 0,
        speed: 0,
        heartRate: 0,
        timestamp: new Date().toISOString()
      };

      this.updateConnectionStatus();
      console.log('Disconnected from all devices');
    } catch (error) {
      console.error('Error during disconnect:', error);
    }
  }

  getCurrentMetrics(): CyclingMetrics {
    return { ...this.currentMetrics };
  }

  getConnectionStatus(): { isConnected: boolean; deviceName: string | null; timestamp: string } {
    return {
      isConnected: this.isConnected,
      deviceName: this.connectedDeviceName,
      timestamp: new Date().toISOString()
    };
  }

  getConnectedDevicesCount(): number {
    return this.connectedDevices.size;
  }
}